src/text-editor-presenter.coffee

```diff
@@ -94,40 +82,68 @@ class TextEditorPresenter
     @updateCustomGutterDecorationState() if @shouldUpdateCustomGutterDecorationState
     @updating = false
 
+    @resetTrackedUpdates()
+
     @state
 
+  resetTrackedUpdates: ->
+    @shouldUpdateFocusedState = false
+    @shouldUpdateHeightState = false
+    @shouldUpdateVerticalScrollState = false
+    @shouldUpdateHorizontalScrollState = false
+    @shouldUpdateScrollbarsState = false
+    @shouldUpdateHiddenInputState = false
+    @shouldUpdateContentState = false
+    @shouldUpdateDecorations = false
+    @shouldUpdateLinesState = false
+    @shouldUpdateCursorsState = false
+    @shouldUpdateOverlaysState = false
+    @shouldUpdateLineNumberGutterState = false
+    @shouldUpdateLineNumbersState = false
+    @shouldUpdateGutterOrderState = false
+    @shouldUpdateCustomGutterDecorationState = false
+
   observeModel: ->
     @disposables.add @model.onDidChange =>
       @updateContentDimensions()
       @updateEndRow()
```

Comment:
Some updates are not deferred until the last possible moment: when a decoration marker changes, for example, we need to add/remove it to/from the decorations cache, e.g.

```coffee
if oldRange.intersectsRowRange(@startRow, @endRow - 1)
  @removeFromLineDecorationCaches(decoration, oldRange)
  intersectsVisibleRowRange = true

if newRange.intersectsRowRange(@startRow, @endRow - 1)
  @addToLineDecorationCaches(decoration, newRange)
  intersectsVisibleRowRange = true
```

I agree that, eventually, we should make (almost) any presenter operation deferred. However, my intent with this PR was to apply behavior-preserving changes just to make the code easier to read. Batching these operations would make the refactoring more invasive and I would feel more confident if we do so when we will have smaller presenters (which will, hopefully, be tested independently).

What do you think?